pkg/operator/controller/externaldns/rbac.go

```diff
@@ -0,0 +1,264 @@
+			Kind:     "ClusterRole",
+			Name:     controller.ExternalDNSBaseName,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      controller.ExternalDNSResourceName(externalDNS),
+				Namespace: namespace,
+			},
+		},
+	}
+}
+
+// currentExternalDNSClusterRole returns true if cluster role exists
+func (r *reconciler) currentExternalDNSClusterRole(ctx context.Context, name types.NamespacedName) (bool, *rbacv1.ClusterRole, error) {
+	cr := &rbacv1.ClusterRole{}
+	if err := r.client.Get(ctx, name, cr); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, cr, nil
+}
+
+// currentExternalDNSClusterRoleBinding returns true if cluster role binding exists
+func (r *reconciler) currentExternalDNSClusterRoleBinding(ctx context.Context, name types.NamespacedName) (bool, *rbacv1.ClusterRoleBinding, error) {
+	crb := &rbacv1.ClusterRoleBinding{}
+	if err := r.client.Get(ctx, name, crb); err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil, nil
+		}
+		return false, nil, err
+	}
+	return true, crb, nil
+}
+
+// createExternalDNSClusterRole creates the given cluster role
+func (r *reconciler) createExternalDNSClusterRole(ctx context.Context, desired *rbacv1.ClusterRole) error {
+	if err := r.client.Create(ctx, desired); err != nil {
+		return fmt.Errorf("failed to create externalDNS cluster role %s: %v", desired.Name, err)
+	}
+	r.log.Info("created externalDNS cluster role", "name", desired.Name)
+	return nil
+}
+
+// createExternalDNSClusterRoleBinding creates the given cluster role binding
+func (r *reconciler) createExternalDNSClusterRoleBinding(ctx context.Context, desired *rbacv1.ClusterRoleBinding) error {
+	if err := r.client.Create(ctx, desired); err != nil {
+		return fmt.Errorf("failed to create externalDNS cluster role binding %s: %v", desired.Name, err)
+	}
+	r.log.Info("created externalDNS cluster role binding", "name", desired.Name)
+	return nil
+}
+
+// updateExternalDNSClusterRole updates the cluster role with the desired state if the rules differ
+func (r *reconciler) updateExternalDNSClusterRole(ctx context.Context, current, desired *rbacv1.ClusterRole) (bool, error) {
+	changed, reason := externalDNSRoleRulesChanged(current.Rules, desired.Rules)
+	if !changed {
+		return false, nil
+	}
+
+	updated := current.DeepCopy()
+	// no complicated resets of specific rules are used
+	// if there is change: reset all
+	updated.Rules = desired.Rules
+	if err := r.client.Update(ctx, updated); err != nil {
+		return false, err
+	}
+	r.log.Info("updated externalDNS cluster role", "name", updated.Name, "reason", reason)
+	return true, nil
+}
+
+// updateExternalDNSClusterRoleBinding updates the cluster role binding with the desired state if the role or subject changed
+func (r *reconciler) updateExternalDNSClusterRoleBinding(ctx context.Context, current, desired *rbacv1.ClusterRoleBinding) (bool, error) {
+	updated := current.DeepCopy()
```

Comment:
Yes, indeed, the pattern is different: not very consistent, I agree but the cases are slightly different.
In case of the cluster role I reset all the rules no matter the change - to avoid the complex resetting of `updated` variable, so it's a single field.
In the cluster role binding case `updated` parameter can be set with the desired values in 3 different fields.

I changed the comment of `externalDNSRoleBindingChanged` to better clarify this.